src/common/tools/gen_listings_and_webworkers.ts

```diff
@@ -0,0 +1,140 @@
+  }
+}
+
+if (argv.length < 4) {
+  usage(0);
+}
+
+const myself = 'src/common/tools/gen_listings_and_webworkers.ts';
+
+const outDir = argv[2];
+
+for (const suiteDir of argv.slice(3)) {
+  // Run concurrently for each suite (might be a tiny bit more efficient)
+  void crawl(suiteDir, false).then(listing => {
+    const suite = path.basename(suiteDir);
+
+    // Write listing.js
+    const outFile = path.normalize(path.join(outDir, `${suite}/listing.js`));
+    fs.mkdirSync(path.join(outDir, suite), { recursive: true });
+    fs.writeFileSync(
+      outFile,
+      `\
+// AUTO-GENERATED - DO NOT EDIT. See ${myself}.
+
+export const listing = ${JSON.stringify(listing, undefined, 2)};
+`
+    );
+
+    // Write suite/webworker/**/*.worker.js
+    for (const entry of listing) {
+      if ('readme' in entry) continue;
+
+      const outFileDir = path.join(
+        outDir,
+        suite,
+        'webworker',
+        ...entry.file.slice(0, entry.file.length - 1)
+      );
+      const outFile = path.join(outDir, suite, 'webworker', ...entry.file) + '.worker.js';
+
+      const relPathToSuiteRoot = Array<string>(entry.file.length).fill('..').join('/');
+
+      fs.mkdirSync(outFileDir, { recursive: true });
+      fs.writeFileSync(
+        outFile,
+        `\
+// AUTO-GENERATED - DO NOT EDIT. See ${myself}.
+
+// oldG is a TestGroup<Fixture> object (defined in common/internal/test_group.ts).
+import { g as oldG } from '${relPathToSuiteRoot}/${entry.file.join('/')}.spec.js';
+
+// FIXME: Expose a proxied test interface. I think this can completely replace test_worker-worker.js
+// (using this instead of that), but if not then hopefully it can at least share code with it.
+console.log(oldG.iterate());
+
+import { globalTestConfig } from '/out/common/framework/test_config.js';
+import { Logger } from '/out/common/internal/logging/logger.js';
+import { setDefaultRequestAdapterOptions } from '/out/common/util/navigator_gpu.js';
+
+async function reportTestResults(ev) {
+  const query = ev.data.query;
+  const expectations = ev.data.expectations;
+  const ctsOptions = ev.data.ctsOptions;
+
+  const { debug, unrollConstEvalLoops, powerPreference, compatibility } = ctsOptions;
+  globalTestConfig.unrollConstEvalLoops = unrollConstEvalLoops;
+  globalTestConfig.compatibility = compatibility;
+
+  Logger.globalDebugMode = debug;
+  const log = new Logger();
+
+  if (powerPreference || compatibility) {
+    setDefaultRequestAdapterOptions({
+      ...(powerPreference && { powerPreference }),
+      // MAINTENANCE_TODO: Change this to whatever the option ends up being
+      ...(compatibility && { compatibilityMode: true }),
+    });
+  }
+
+  // const testcases = Array.from(await loader.loadCases(parseQuery(query)));
+  // assert(testcases.length === 1, 'worker query resulted in != 1 cases');
+
+  // const testcase = testcases[0];
+  const testcase = { query }; // FIXME! I failed to figure out how to get a testcase from oldG and query ;(
```

Comment:
We've never done this before so it'll require work. I'm not quite sure.

I think you can cast oldG to `TestGroup<Fixture>` to get at its internals.
If your query is a string, use `parseQuery` to parse it to get the [`testParts`](url) and [`params`](https://github.com/gpuweb/cts/blob/62389bc770bb9d86670881dadc24a8ea8e259d3e/src/common/internal/query/query.ts#L105) from it.
Use the testParts to search for the test in the TestGroup.
Then you can probably just call the test's function with the params?
loadCases wraps this up normally, you can look at how it does it.

(Aside: I realized it doesn't need to be called `oldG`. There will be no new `g`. So you can just change the code I wrote to `import { g } ...` I think.

Reply:
I've tried this code without success.
I believe I'm missing instantiating the test but it's not clear where it's done.

```js
  try {
    const testQuerySingleCase = parseQuery(query);
    console.log({testQuerySingleCase});
    let testcase = null;
    for (const t of oldG.iterate()) {
      for (const c of t.iterate(testQuerySingleCase.params)) {
        if (comparePublicParamsPaths(c.id.params, testQuerySingleCase.params) === Ordering.Equal) {
          testcase = c;
        }
      }
    }
    console.log({testcase});
    testcase.fn();
    console.log('yeah');
  } catch(error) {
    console.log(error);
  }
```

![image](https://github.com/gpuweb/cts/assets/634478/9be1999b-724b-4b10-aa69-b0c4f62db31b)

Reply:
Argh... I thought I was cracking it with the following code but I reached a WebSocket error `WebSocket connection to 'ws://localhost:59497/optional_cts_websocket_logger' failed: Error in connection establishment: net::ERR_CONNECTION_REFUSED`

```js
  const testQuerySingleCase = parseQuery(query);
  let testcase = null;
  for (const t of oldG.iterate()) {
    for (const c of t.iterate(testQuerySingleCase.params)) {
      if (comparePublicParamsPaths(c.id.params, testQuerySingleCase.params) === Ordering.Equal) {
        testcase = c;
      }
    }
  }
  const [rec, result] = log.record(query);
  await testcase.run(rec, expectations);
  this.postMessage({ query, result });
```

![image](https://github.com/gpuweb/cts/assets/634478/d20375dd-0a33-4637-b945-b01f815d64ed)

Reply:
It turns out the WebSocket error is not the blocker issue.
The fail now tests (properly?) with the following surprising error as `expectations` is an Array of size 0.

```js
  const [rec, result] = log.record(query);
  console.log({expectations}); // <----
  await testcase.run(rec, expectations);
  this.postMessage({ query, result });
}
```

```
Error: expectations is not iterable
TypeError: expectations is not iterable
    at getExpectedStatus (http://localhost:8080/out/common/internal/test_group.js:561:25)
    at RunCaseSpecific.run (http://localhost:8080/out/common/internal/test_group.js:717:13)
    at async WindowClient.reportTestResults (http://localhost:8080/out/webgpu/webworker/api/operation/adapter/requestAdapter.worker.js:47:3)
```

![image](https://github.com/gpuweb/cts/assets/634478/4d2bdb6c-64da-4395-b0fd-2dc493f2393d)

Reply:
Using the following code allows me to run a small subset of tests (http://localhost:8080/standalone/?worker=service&q=webgpu:api,operation,adapter,requestAdapter:requestAdapter:*), which is a small victory (yeah!):

```js
  const testQuerySingleCase = parseQuery(query);
  let testcase = null;
  for (const t of g.iterate()) {
    for (const c of t.iterate(testQuerySingleCase.params)) {
      if (comparePublicParamsPaths(c.id.params, testQuerySingleCase.params) === Ordering.Equal) {
        testcase = c;
      }
    }
  }
  const [rec, result] = log.record(query);
  await testcase.run(rec, testQuerySingleCase, expectations);
  this.postMessage({ query, result });
  ```

The problem now comes from Service Worker I believe.
When running many tests at once (http://localhost:8080/standalone/?worker=service&q=webgpu:api,operation,*), the service worker file I get is not the one I expect. More specifically, the following example shows I got requestAdapter.worker.js instead of requestAdapterInfo.worker.js.
It's not clear to me what is happening but the creation of many service worker files may not help ;(

![image](https://github.com/gpuweb/cts/assets/634478/72624ad4-7ebe-45f0-ba43-8343b1225c37)


```js

    const registration = await navigator.serviceWorker.register(serviceWorkerPath, {
      type: 'module',
      scope: '/',
    });
    await navigator.serviceWorker.ready;
    if (registration.active.scriptURL !== `${window.location.origin}${serviceWorkerPath}`) {
      console.error(`Failed to retrieve proper file from service worker registration.
          Got ${registration.active.scriptURL} instead of ${window.location.origin}${serviceWorkerPath}`);
      return;
    }
 ```

Reply:
👋  I finally got it working!